File: Src/ECScape.Core/Systems/SpawnerSystem.cs
using ECScape.Core.Components;
using ECScape.Core.Engine;
using ECScape.Core.Entities;

namespace ECScape.Core.Systems;

internal class SpawnerSystem : ISystem
{
    private DateTime LastMeteroid = DateTime.Now;
    private DateTime LastPowerUp = DateTime.Now;
    private DateTime LastMeteroidRoll = DateTime.Now;
    private DateTime LastPowerUpRoll = DateTime.Now;
    private static readonly TimeSpan MinTimeBetween = TimeSpan.FromSeconds(4);
    private static readonly TimeSpan TimeBetweenRolls = TimeSpan.FromSeconds(1);

    public void Update(World world, float _)
    {
        SpawnCollectables(world);
        SpawnMeteroid(world);
        SpawnPowerUp(world);
    }

    private void SpawnMeteroid(World world)
    {
        if (DateTime.Now - LastMeteroid < MinTimeBetween) { return; }
        if (DateTime.Now - LastMeteroidRoll < TimeBetweenRolls) { return; }
        LastMeteroidRoll = DateTime.Now;
        if (World.Random.Next(0, 10) < 2) { return; }
        EntityFactory.CreateMeteoroid(world, World.Random.Next(0, UiInterface.WorldWidth + 1), 1, 1);
        LastMeteroid = DateTime.Now;
    }

    private void SpawnPowerUp(World world)
    {
        if (DateTime.Now - LastPowerUp < MinTimeBetween) { return; }
        if (DateTime.Now - LastPowerUpRoll < TimeBetweenRolls) { return; }
        LastPowerUpRoll = DateTime.Now;
        if (World.Random.Next(0, 10) < 3) { return; }
        EntityFactory.CreatePowerUp(world, World.Random.Next(0, UiInterface.WorldWidth + 1), 1, 1);
        LastPowerUp = DateTime.Now;
    }

    private static void SpawnCollectables(World world)
    {
        var collectables = world.GetEntitiesWith(typeof(Collectable), typeof(Exists));
        if (collectables.Count() >= 2) { return; }

        var width = World.Random.Next(1, 5);
        var height = World.Random.Next(1, 5);
        EntityFactory.CreateCollectable(
            world,
            World.Random.Next(0, UiInterface.WorldWidth + 1),
            World.Random.Next(0, UiInterface.WorldHeight + 1),
            World.Random.Next(0, 2) == 0 ? '$' : '♥',
            width,
            height,
            Math.Min(width, height));
    }
}

File: src/systems/SpawnerSystem.js
import ISystem from './ISystem.js';
import EntityFactory from '../entities/EntityFactory.js';
import { Collectable, Exists } from '../components/Components.js';
import UiInterface from '../core/uiInterface.js';

class SpawnerSystem extends ISystem {
    constructor() {
        super();
        this.lastMeteoroid = Date.now();
        this.lastPowerUp = Date.now();
        this.lastMeteoroidRoll = Date.now();
        this.lastPowerUpRoll = Date.now();
        this.minTimeBetween = 4000;
        this.timeBetweenRolls = 1000;
    }

    update(world, deltaTime) {
        this.spawnCollectables(world);
        this.spawnMeteoroid(world);
        this.spawnPowerUp(world);
    }

    spawnMeteoroid(world) {
        if (Date.now() - this.lastMeteoroid < this.minTimeBetween) return;
        if (Date.now() - this.lastMeteoroidRoll < this.timeBetweenRolls || Math.random() < 0.2) {
            this.lastMeteoroidRoll = Date.now();
            return;
        }
        EntityFactory.createMeteoroid(world, Math.floor(Math.random() * (UiInterface.WorldWidth + 1)), 1, 1);
        this.lastMeteoroid = Date.now();
    }

    spawnPowerUp(world) {
        if (Date.now() - this.lastPowerUp < this.minTimeBetween) return;
        if (Date.now() - this.lastPowerUpRoll < this.timeBetweenRolls || Math.random() < 0.3) {
            this.lastPowerUpRoll = Date.now();
            return;
        }
        EntityFactory.createPowerUp(world, Math.floor(Math.random() * (UiInterface.WorldWidth + 1)), 1, 1);
        this.lastPowerUp = Date.now();
    }

    spawnCollectables(world) {
        const collectables = world.getEntitiesWith(Collectable, Exists);
        if (collectables.length >= 2) return;

        const width = Math.floor(1 + Math.random() * 4);
        const height = Math.floor(1 + Math.random() * 4);
        EntityFactory.createCollectable(
            world,
            Math.floor(Math.random() * (UiInterface.WorldWidth + 1)),
            Math.floor(Math.random() * (UiInterface.WorldHeight + 1)),
            Math.random() < 0.5 ? '$' : '?',
            width,
            height,
            Math.min(width, height)
        );
    }
}

export default SpawnerSystem;